File: execution/organize_skills.py
import os
import shutil
import re
import argparse

def organize_skills(repo_root, target_root):
    catalog_path = os.path.join(repo_root, "CATALOG.md")
    skills_src = os.path.join(repo_root, "skills")
    
    if not os.path.exists(catalog_path):
        print(f"Error: Catalog not found at {catalog_path}")
        return

    if not os.path.exists(skills_src):
        print(f"Error: Skills source folder not found at {skills_src}")
        return

    # Ensure target_root exists
    if not os.path.exists(target_root):
        os.makedirs(target_root)

    with open(catalog_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Split by H2 headers (categories)
    # Pattern: ## category_name (count)
    categories = re.split(r'^##\s+', content, flags=re.MULTILINE)
    
    for cat_block in categories[1:]: # Skip the first part before the first category
        lines = cat_block.split('\n')
        category_header = lines[0].strip()
        # Extract category name (remove (count) if present)
        category_name = re.sub(r'\s*\(\d+\)$', '', category_header).strip()
        category_name = "".join([c if c.isalnum() or c in (' ', '-', '_') else '_' for c in category_name])
        
        category_path = os.path.join(target_root, category_name)
        if not os.path.exists(category_path):
            os.makedirs(category_path)
            
        print(f"Processing category: {category_name}")
        
        # Skill names are in the first column of the table, wrapped in backticks
        # Example: | `skill-name` | Description |
        skill_names = re.findall(r'^\|\s*`([^`]+)`\s*\|', cat_block, flags=re.MULTILINE)
        
        for skill in skill_names:
            src_skill_path = os.path.join(skills_src, skill)
            dst_skill_path = os.path.join(category_path, skill)
            
            if os.path.exists(src_skill_path):
                try:
                    # Use shutil.copytree if the skill is a directory
                    if os.path.isdir(src_skill_path):
                        if os.path.exists(dst_skill_path):
                             shutil.rmtree(dst_skill_path)
                        shutil.copytree(src_skill_path, dst_skill_path)
                    else:
                        shutil.copy2(src_skill_path, dst_skill_path)
                except Exception as e:
                    print(f"  Error copying {skill}: {e}")
            else:
                print(f"  Warning: Skill {skill} not found in {skills_src}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Organize skills by category.')
    parser.add_argument('--repo', required=True, help='Path to the cloned repository root.')
    parser.add_argument('--target', required=True, help='Path where organized skills will be saved.')
    
    args = parser.parse_args()
    organize_skills(args.repo, args.target)

File: execution/organize_skills.js
const fs = require('fs');
const path = require('path');

function organizeSkills(repoRoot, targetRoot) {
    const catalogPath = path.join(repoRoot, 'CATALOG.md');
    const skillsSrc = path.join(repoRoot, 'skills');

    if (!fs.existsSync(catalogPath)) {
        console.error(`Error: Catalog not found at ${catalogPath}`);
        return;
    }

    if (!fs.existsSync(skillsSrc)) {
        console.error(`Error: Skills source folder not found at ${skillsSrc}`);
        return;
    }

    if (!fs.existsSync(targetRoot)) {
        fs.mkdirSync(targetRoot, { recursive: true });
    }

    const content = fs.readFileSync(catalogPath, 'utf8');
    // Split by categories: ## category_name (count)
    const categoryBlocks = content.split(/^##\s+/m).slice(1);

    categoryBlocks.forEach(block => {
        const lines = block.split('\n');
        const header = lines[0].trim();
        // Remove (count) if exists
        let categoryName = header.replace(/\s*\(\d+\)$/, '').trim();
        // Sanitize category name
        categoryName = categoryName.replace(/[^\w\s-]/g, '_');

        const categoryPath = path.join(targetRoot, categoryName);
        if (!fs.existsSync(categoryPath)) {
            fs.mkdirSync(categoryPath, { recursive: true });
        }

        console.log(`Processing category: ${categoryName}`);

        // Skill names in backticks in the first column
        const skillRegex = /^\|\s*`([^`]+)`\s*\|/gm;
        let match;
        while ((match = skillRegex.exec(block)) !== null) {
            const skill = match[1];
            const srcPath = path.join(skillsSrc, skill);
            const dstPath = path.join(categoryPath, skill);

            if (fs.existsSync(srcPath)) {
                try {
                    copyRecursive(srcPath, dstPath);
                } catch (err) {
                    console.error(`  Error copying ${skill}: ${err.message}`);
                }
            } else {
                console.warn(`  Warning: Skill ${skill} not found in ${skillsSrc}`);
            }
        }
    });
}

function copyRecursive(src, dest) {
    const stats = fs.statSync(src);
    if (stats.isDirectory()) {
        if (!fs.existsSync(dest)) {
            fs.mkdirSync(dest, { recursive: true });
        }
        fs.readdirSync(src).forEach(child => {
            copyRecursive(path.join(src, child), path.join(dest, child));
        });
    } else {
        fs.copyFileSync(src, dest);
    }
}

const args = process.argv.slice(2);
if (args.length < 2) {
    console.log("Usage: node organize_skills.js <repo_path> <target_path>");
} else {
    organizeSkills(args[0], args[1]);
}
